src/ggml-metal.metal

```diff
@@ -487,6 +488,7 @@ kernel void kernel_soft_max_4(
     }
 
     const float lsum = lsum4[0] + lsum4[1] + lsum4[2] + lsum4[3];
+    threadgroup_barrier(mem_flags::mem_threadgroup);
```

Comment:
Do these barriers make a difference in your tests?
To me they seems superfluous, but I could be missing something

Reply:
I am not sure why, but without the barrier this test case fails sometimes:
```C++
    test_cases.emplace_back(new test_soft_max(GGML_TYPE_F32, {4096, 4096, 8, 1}));
```

Reply:
Using a execution barrier only `threadgroup_barrier(mem_flags::mem_none);` seems to be enough to fix the issue. My intuition when adding the barrier is to ensure that all threads finished computing their value before the `simd_sum`, but I don't know enough about Metal to tell if that's actually require. In CUDA the warp shuffle functions have an implicit synchronization, so it is not necessary.